x/superfluid/keeper/unpool.go

```diff
@@ -0,0 +1,108 @@
+package keeper
+
+import (
+	"time"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	gammtypes "github.com/osmosis-labs/osmosis/v7/x/gamm/types"
+
+	"github.com/osmosis-labs/osmosis/v7/x/superfluid/types"
+)
+
+func (k Keeper) UnpoolAllowedPools(ctx sdk.Context, sender sdk.AccAddress, poolId uint64, lockId uint64) (uint64, error) {
+	// check if pool is whitelisted for unpool
+	allowedPools := k.GetUnpoolAllowedPools(ctx)
+	allowed := false
+
+	for _, allowedPoolId := range allowedPools {
+		if poolId == allowedPoolId {
+			allowed = true
+		}
+	}
+
+	if !allowed {
+		return 0, types.ErrUnpoolNotAllowed
+	}
+
+	// check if lock is unlocking
+	lock, err := k.lk.GetLockByID(ctx, lockId)
+	if err != nil {
+		return 0, err
+	}
+
+	gammShare := lock.Coins[0]
+	if gammShare.Denom != gammtypes.GetPoolShareDenom(poolId) {
+		return 0, types.ErrLockUnpoolNotAllowed
+	}
+
```

Comment:
```suggestion
// Steps for unpooling
// 1) If superfluid delegated, superfluid undelegate
// 2) Break underlying lock. This will clear any metadata if things are superfluid unbonding
// 3) Get duration from {} (Consider if we can handle complexity for already unbonding Locks)
// 4) ExitPool with these unlocked LP shares
// 5) Make 1 new lock for every asset in collateral. Many code paths need this assumption to hold
// 6) Make new lock begin unlocking
```

Reply:
Then clarify / perhaps document each component further. Worth considering splitting them into separate functions

Reply:
Tysm for this!